src/Dictionaries/TrieDictionary.h

```diff
@@ -207,32 +217,46 @@ class TrieDictionary final : public IDictionaryBase
 
     Attribute createAttributeWithType(const AttributeUnderlyingType type, const Field & null_value);
 
+    template <typename AttributeType, typename OutputType, typename ValueSetter, typename DefaultGetter>
+    void getItemsByTwoKeyColumnsImpl(
+        const Attribute & attribute, const Columns & key_columns, ValueSetter && set_value, DefaultGetter && get_default) const;
 
     template <typename AttributeType, typename OutputType, typename ValueSetter, typename DefaultGetter>
     void
     getItemsImpl(const Attribute & attribute, const Columns & key_columns, ValueSetter && set_value, DefaultGetter && get_default) const;
 
-
     template <typename T>
-    bool setAttributeValueImpl(Attribute & attribute, const StringRef key, const T value);
+    void setAttributeValueImpl(Attribute & attribute, const T value);
 
-    bool setAttributeValue(Attribute & attribute, const StringRef key, const Field & value);
+    void setAttributeValue(Attribute & attribute, const Field & value);
 
     const Attribute & getAttribute(const std::string & attribute_name) const;
 
     template <typename T>
     void has(const Attribute & attribute, const Columns & key_columns, PaddedPODArray<UInt8> & out) const;
 
     Columns getKeyColumns() const;
+    RowIdxConstIter ipNotFound() const;
+    RowIdxConstIter tryLookupIPv4(UInt32 addr, uint8_t * buf) const;
+    RowIdxConstIter tryLookupIPv6(const uint8_t * addr) const;
+
+    template <typename IPContainerType,
+        typename IPValueType = typename IPContainerToValueType<IPContainerType>::value>
+    RowIdxConstIter lookupIP(IPValueType target) const;
+
+    static const uint8_t * getIPv6FromOffset(const IPv6Container & ipv6_col, size_t i);
 
     const DictionaryStructure dict_struct;
     const DictionarySourcePtr source_ptr;
     const DictionaryLifetime dict_lifetime;
     const bool require_nonempty;
     const std::string key_description{dict_struct.getKeyDescription()};
 
+    std::variant<IPv4Container, IPv6Container> ip_column;
+    IPMaskContainer mask_column;
+    ContainerType<size_t> parent_subnet;
```

Comment:
I definitely need comment which explains what is stored here.

Probably, general description about how our keys are stored.